### PR TITLE
Datastore docs

### PR DESCRIPTION
**Pull request by Eunovo**
### Description

This PR is a work in progress. Setup new docs structure. Started Datastore docs.

##### Checklist

- [x] documentation is changed or added

### CODE REVIEW

**Comment by wtrocki** [comment]
Amazing work so far
Can we get screenshot of side panel or list items as tasks?

**Comment by Eunovo** [comment]
image

@wtrocki The docs here is outdated. We can merge this once we approve the structure. I can create another PR later when our API is more stable.

**Comment by wtrocki** [comment]
@Eunovo So suggestion is to lang this to master (with some typo checks etc.) and then extend it from there right ?

File: docs/datastore/cli.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
+---
+id: cli
+title: Datastore CLI
+sidebar_label: Datastore CLI
+---
+
+## Using the CLI tool
+
+We provide a graphback plugin to generate required config for datastore.
```
Comment (wtrocki) [inline docs/datastore/cli.md:9]: If I put myself as newbie I will really struggle to get sense from this. 
There is strong assumption that someone will have graphback installed. 

There is no single cli or instruction to get that done. 
Not sure if that is problem with doc, but flagging this as docs here are really confusing and we simply missing info in order to replicate this from beginning.

File: docs/datastore/crud-operations.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+TaskModel.query({ title: "test" })
+.then((data) => {}) // Retrieves all tasks where title matches "test"
+```
+
+## Updating data
+
+```typescript
+import { TaskModel } from './datastoreConfig';
+
+// update the title of all tasks with title = "test" to "Offix Test"
```
Comment (wtrocki) [inline docs/datastore/crud-operations.md:42]: This case is rare! We should show update by id. IMHO updating multiple items in client should not be even possible (it brings serious risk and makes it hard to implement security)

File: docs/datastore/crud-operations.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+})
+```
+
+## Deleting data
+
+```typescript
+import { TaskModel } from './datastoreConfig';
+
+// delete all tasks with numberOfDaysLeft greater than 4
+await TaskModel.remove({ numberOfDaysLeft: { gt: 4 } });
```
Comment (wtrocki) [inline docs/datastore/crud-operations.md:57]: Same as above. This is not about deleting data. Most of the people will be looking to operate on their models and then replicate change - involves using id. 

We need to update this and possible having separate chapter (batch updates) or something.

File: docs/datastore/datastore-api.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
+---
```
Comment (wtrocki) [inline docs/datastore/datastore-api.md:1]: IMHO this should be generated from typedoc. We can keep this but it will outdate really fast.

File: docs/datastore/datastore-api.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+Configuration options for `DataStore`.
+
+| Input | Type | Description |
+| ----- | ---- | ----------- |
+| dbName | ?string | The Database name |
+| schemaVersion | ?number | The Schema Version number. Used to trigger a Schema upgrade. Defaults to 1 |
+| replicationConfig | ?[GlobalReplicationConfig](replication-api#GlobalReplicationConfig) | Configuration for Replication engine |
+
+### CustomEngines
```
Comment (wtrocki) [inline docs/datastore/datastore-api.md:25]: This is internal - it will be covered in the platform specific config (react native etc.)

File: docs/datastore/filters.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
+---
+id: filters
+title: Filtering
+sidebar_label: Filtering
+---
+
+## Filter Format
```
Comment (wtrocki) [inline docs/datastore/filters.md:7]: Lack of introduction and general info. 
What we filtering. How to use filters.

File: docs/datastore/introduction.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
+---
+id: introduction
+title: Introduction to Offix Datastore
+sidebar_label: Offix Datastore
+---
+
+Welcome to the Offix Datastore guide! Offix Datastore is a database for
```
Comment (wtrocki) [inline docs/datastore/introduction.md:7]:
```suggestion
Offix Datastore is an fully capable client side store  for
```

File: docs/datastore/introduction.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+JavaScript apps with real-time synchronization over GraphQL.
+Offix Datastore saves all your data locally while pushing them your server,
```
Comment (wtrocki) [inline docs/datastore/introduction.md:9]:
```suggestion
Offix Datastore saves all your data locally and enables GraphQL replication mechanism to your generic GraphQL server,
```

File: docs/datastore/introduction.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+Offix Datastore allows you to build "Offline-first" apps with ease.
+User data is always available locally and local database is
+synchronized with server as soon as the user is online.
+
+#### GraphQL CRUD compatible
+
+Offix Datastore supports synchronization with [GraphQL CRUD](https://graphqlcrud.org/) compatible servers by default.
+If your server doesn't support GraphQLCrud, you can build your synchronization engine.
+[TODO link how to build replication engine here]
```
Comment (wtrocki) [inline docs/datastore/introduction.md:25]:
```suggestion
```

File: docs/datastore/model-api.md [hunk 10]
```diff
@@ -0,0 +1,62 @@
+---
```
Comment (wtrocki) [inline docs/datastore/model-api.md:1]: Seems like we can merge this information with stuff above have single chapter.
WDYT?

File: docs/datastore/model-api.md [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+
+### model.subscribe
+
+Accepts a [CRUDEvent](storage-api#CRDUEvents) and a callback.
+The callback is called with a [StoreChangeEvent](storage-api#StoreChangeEvent)
+when the specified `CRUDEvent` is fired.
+
+
+## Model JSON Schema
```
Comment (wtrocki) [inline docs/datastore/model-api.md:40]: This is probably true reference of the model API but really internal and will outdated faster than sandwitch. In fact current schema is already outdated. 

Let avoid adding any internal stuff for the moment. We can just document react use cases and some operations.

File: docs/datastore/model-api.md [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+| namespace | ?string | TODO |
+| keyPath | ?string | TODO |
+| version | ?number | TODO |
+| indexes | ?string[] | TODO |
+| encrypted | ?string[] | TODO |
+| primaryKey | string | TODO |
+| type | "object" | TODO |
+| properties | [Fields](#Fields) | TODO |
+
+### Fields
```
Comment (wtrocki) [inline docs/datastore/model-api.md:54]: same as above.

File: docs/datastore/react.md [hunk 13]
```diff
@@ -0,0 +1,179 @@
+---
+id: react
+title: React - using Datastore Hooks
+sidebar_label: React
+---
+
+Offix Datastore provides hooks for CRUD operations and subscription.
```
Comment (wtrocki) [inline docs/datastore/react.md:7]:
```suggestion
Offix Datastore provides react hooks for CRUD operations and subscription.
```

File: docs/datastore/react.md [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+
+For a quick start see [sample react app](https://github.com/aerogear/offix/tree/master/examples/react-datastore).
+
+## useSave
+
+The `useSave` hook provides a lazy `save` function,
```
Comment (wtrocki) [inline docs/datastore/react.md:13]: Question - how useSave will work with saveById?

File: docs/datastore/react.md [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+
+const AddTask = () => {
+    const [taskDetails, setTaskDetails] = useState({
+        title: "", description: "", numberOfDaysLeft: 0
+    });
+    const { isLoading, error, save } = useSave(TaskModel);
+
+    async function handleSubmit(e) {
+        e.preventDefault();
+        const result = await save(taskDetails);
```
Comment (wtrocki) [inline docs/datastore/react.md:29]: Question - we will dynamically dectect id and use it to save?

File: docs/datastore/react.md [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+    }
+    
+    return ...
+}
+```
+
+## useSubscription
+
+You can subscribe to specific events and receive changes data using this hook.
+Below, we listen for updates to the task and render them. The `data` returned
```
Comment (wtrocki) [inline docs/datastore/react.md:168]:
```suggestion
We listen for updates to the task and render them. The `data` returned
```

File: docs/datastore/react.md [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+by `useSubscription` is the data carried by the change event.
+
+```javascript
+const Task = ({ task }) => {
+    // Listen for updates on this task
+    const { data } = useSubscription(TaskModel, CRUDEvents.UPDATE, task);
```
Comment (wtrocki) [inline docs/datastore/react.md:174]: Question people might have: Can I do `CRUDEvents.UPDATE | CRUDEvents.DELETE`

File: docs/datastore/ref-getting-started.md [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+
+Or [yarn](https://yarnpkg.com/en/package/offix-datastore):
+
+```shell
+yarn add offix-datastore
+```
+
+## Using DataStore
+
+### Sample GraphQL Schema
```
Comment (wtrocki) [inline docs/datastore/ref-getting-started.md:23]:
```suggestion
### Building your own GraphQL type
```

File: docs/datastore/ref-getting-started.md [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+
+We have a [cli tool](cli.md) that generates DataStore config and Model JSON schema given a GraphQL schema.
+`DataStore` accepts a config object and an optional `CustomEngines`, see [api reference](datastore-api).
+
+### Datastore Models
+
+To be able to store user tasks in DataStore, you need to create its DataStore model.
+The DataStore model provides the API to perform CRUD operations on `Task` in the DataStore.
+
+Let's create the Task DataStore model.
```
Comment (wtrocki) [inline docs/datastore/ref-getting-started.md:47]: This is oudated. We should create guide how to generate this instead of creating this by hand :)

File: docs/datastore/ref-getting-started.md [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+import schema from "./schema.json";
+
+const datastore = new DataStore({
+  dbName: "offix-datastore",
+  replicationConfig: {
+    client: {
+      url: "http://localhost:4000/graphql",
+      wsUrl: "ws://localhost:4000/graphql",
+    },
+    delta: { enabled: true },
```
Comment (wtrocki) [inline docs/datastore/ref-getting-started.md:107]: This is redundant - we should suggest defaults.

File: docs/datastore/ref-getting-started.md [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+
+To make DataStore acknowledge these changes, you need to increment the schema version.
+
+Using our sample app, suppose we add a `SubTask` model, we need to increment 
+the schema version to trigger the creation of the "user_SubTask" table on the client device.
+
+The `DataStore` constructor takes a schema version parameter(defaults to 1). 
+
+```typescript
+const dataStore = new DataStore({
```
Comment (wtrocki) [inline docs/datastore/ref-getting-started.md:137]: This can be handled by cli as well.

File: docs/datastore/cli.md (hunk 1 above)
Comment (Eunovo) [reply docs/datastore/cli.md:9]: I haven't updated this

**Comment by wtrocki** [comment]
TL;DR - Less is better for the moment. I have pointed some of the issues that prevent us from merging.

File: docs/datastore/datastore-api.md (hunk 4 above)
Comment (Eunovo) [reply docs/datastore/datastore-api.md:1]: Can we? I know users could generate from javadoc in Java. I will check for tools to help with this

File: docs/datastore/introduction.md (hunk 9 above)
Comment (Eunovo) [reply docs/datastore/introduction.md:25]: Are we not going to show how to build custom replication engine?

File: docs/datastore/react.md (hunk 17 above)
Comment (Eunovo) [reply docs/datastore/react.md:174]: This seems like syntax from Java. I don't think it's possible in JS. To replicate this feature we can accept an array of CRUDEvents

File: docs/datastore/ref-getting-started.md (hunk 19 above)
Comment (Eunovo) [reply docs/datastore/ref-getting-started.md:47]: I thought the manual guide would still be of use even with the cli guide. We can take out this part.

File: docs/datastore/ref-getting-started.md (hunk 20 above)
Comment (Eunovo) [reply docs/datastore/ref-getting-started.md:107]: How do you mean?

File: docs/datastore/model-api.md (hunk 10 above)
Comment (Eunovo) [reply docs/datastore/model-api.md:1]: I think this stuff belongs to API reference section. That's how I see it.

File: docs/datastore/crud-operations.md (hunk 2 above)
Comment (Eunovo) [reply docs/datastore/crud-operations.md:42]: This doc is outdated. It hasn't been updated to include new API that has "ById" methods.  Also, why does this make it hard to implement security? Security is done on the server and we don't push batch updates to server.

File: docs/datastore/react.md (hunk 14 above)
Comment (Eunovo) [reply docs/datastore/react.md:13]: I think we will need new hooks. WDYT?

File: docs/datastore/model-api.md (hunk 10 above)
Comment (wtrocki) [reply docs/datastore/model-api.md:1]: I would not have api reference yet. Model docs are really basic now so I was thinking to add more stuff there. Reference makes no sense if you have alpha version of the project.

File: docs/datastore/react.md (hunk 14 above)
Comment (wtrocki) [reply docs/datastore/react.md:13]: definetly need new issue. We cah use something like `upsert:true` 

And the way things are updated now is really strange you can pass array (does it mean you want to remove all objects). I think our hooks should be by id only for the moment, but need to check how we specify id (passing entire object vs id + effect)

File: docs/datastore/react.md (hunk 17 above)
Comment (wtrocki) [reply docs/datastore/react.md:174]: I meant that this is inline function that returns true. Array will work as well.

File: docs/datastore/ref-getting-started.md (hunk 20 above)
Comment (wtrocki) [reply docs/datastore/ref-getting-started.md:107]: Remove all items bellow

File: docs/datastore/react.md (hunk 14 above)
Comment (Eunovo) [reply docs/datastore/react.md:13]: Where do we pass array? Our api doesn't accept arrays anywhere(that I know of). I think It's okay for us to restrict update and delete to by Id. Also, the way we specify id doesn't matter much IMO. Passing entire object could be better if you are look at how users send updates to API. 
If you support passing entire object and restricting update and delete to by id, I can push this in new PR. Also, having and `upsert` option for `useUpdate` hook seems like the best approach right now. Since, you approve, I can push that too.

Comment (wtrocki) [reply docs/datastore/react.md:13]: Let's create issue to chat about it, but feel free to get upset as quick win

Comment (wtrocki) [reply docs/datastore/react.md:13]: Let's create issue to chat about it, but feel free to get upset as quick win

**Comment by wtrocki** [comment]
We can publish docs without using version: https://offix.dev/docs/next/getting-started
And use this for initial blog post and other work.

**Comment by wtrocki** [comment]
@Eunovo Ping me when you are readyu for another review. 
Deployed (using GIT_USER=wtrocki USE_SSH=true yarn deploy)

**Comment by wtrocki** [comment]
Ok. Added this change manualy. See: https://offix.dev/docs/next/offix/getting-started

File: docs/datastore/cli.md (hunk 1 above)
Comment (Eunovo) [reply docs/datastore/cli.md:9]: We don't have an offix cli. When we discussed clis, you said it would be difficult to manage two clis. I don't know how I'm going to spin this. They have to install graphback. I already point to graphback docs. I will add small Graphback intro but I don't think it will make much difference since they still have to go to the graphback docs.

Comment (wtrocki) [reply docs/datastore/cli.md:7]: I have commented that while ago. We kinda need small wrapper around graphback that will create config in memory rather than in the project. This will simplify docs.

Comment (Eunovo) [reply docs/datastore/cli.md:7]: Does the Graphback cli support such customization?

File: docs/datastore/ref-getting-started.md (hunk 19 above)
Comment (wtrocki) [reply docs/datastore/ref-getting-started.md]: CLI is esentiall. Telling people to do tons of boilerplate makes no sense.

File: docs/datastore/introduction.md (hunk 9 above)
Comment (wtrocki) [reply docs/datastore/introduction.md]: We going to do it, but not in introduction!

File: docs/datastore/datastore-api.md (hunk 4 above)
Comment (wtrocki) [reply docs/datastore/datastore-api.md]: Do not check/do stuff without issue. There are tools like typedoc. We can do that in no time. 
Just do not do stuff manually.

File: docs/datastore/cli.md (hunk 1 above)
Comment (wtrocki) [reply docs/datastore/cli.md:7]: Yep. It is programatical usage vs cli/config (we support both)
Docs shows us missing pieces so.

https://github.com/aerogear/offix/issues/729

**Comment by wtrocki** [comment]
Works for me!
